pkg/corerp/api/v20231001preview/environment_conversion.go

```diff
@@ -408,35 +420,118 @@ func fromRecipePropertiesClassificationDatamodel(e datamodel.EnvironmentRecipePr
+
+			// Account for possibility of nested additional properties during unmarshalling
+			// https://json-schema.org/understanding-json-schema/reference/object#additionalproperties
+			if tmpAddnlProperties, ok := addnlProperties["additionalProperties"]; ok && addnlProperties["additionalProperties"] != nil {
+				if tmpMap, ok := tmpAddnlProperties.(map[string]any); ok {
+					addnlProperties = tmpMap
+				} else {
+					return nil, v1.ErrInvalidModelConversion
+				}
+			}
+
+			// check if additional properties contain Secrets
+			// get AdditionalProperties from providerAdditionalProperties excluding the 'secrets' key.
+			for key, value := range addnlProperties {
+				if key == "secrets" {
+					if tmpSecret, ok := value.(map[string]*SecretReference); !ok {
+						return nil, v1.ErrInvalidModelConversion
+					} else {
+						propSecrets := getSecretsFromProviderProperties(tmpSecret)
+
+						// Add the secrets to the convertedSecrets map
+						for secretKey, secretValue := range propSecrets {
+							if convertedSecrets == nil {
+								convertedSecrets = map[string]datamodel.SecretReference{}
+							}
+							convertedSecrets[secretKey] = secretValue
+						}
+					}
+
+					// remove the 'secrets' key from the additional properties as we've added all
+					// secrets to convertedSecrets map
+					delete(addnlProperties, "secrets")
+				}
+			}
+
 			dm[k] = append(dm[k], datamodel.ProviderConfigProperties{
-				AdditionalProperties: providerAdditionalProperties,
+				AdditionalProperties: addnlProperties,
+				Secrets:              convertedSecrets,
 			})
 		}
 	}
 
-	return dm
+	return dm, nil
+}
+
+func getSecretsFromProviderProperties(providerProperties map[string]*SecretReference) map[string]datamodel.SecretReference {
+	var convertedSecrets map[string]datamodel.SecretReference
+
+	for secretKey, secretValue := range providerProperties {
+		if secretValue != nil {
+			convertedSecret := datamodel.SecretReference{
+				Source: to.String(secretValue.Source),
+				Key:    to.String(secretValue.Key),
+			}
+
+			if convertedSecrets == nil {
+				convertedSecrets = map[string]datamodel.SecretReference{}
+			}
+			convertedSecrets[secretKey] = convertedSecret
+		}
+	}
+
+	return convertedSecrets
 }
 
-func fromRecipeConfigTerraformProvidersDatamodel(config datamodel.RecipeConfigProperties) map[string][]map[string]any {
+func fromRecipeConfigTerraformProvidersDatamodel(config datamodel.RecipeConfigProperties) map[string][]*ProviderConfigProperties {
 	if config.Terraform.Providers == nil {
 		return nil
 	}
 
-	providers := map[string][]map[string]any{}
+	providers := map[string][]*ProviderConfigProperties{}
 	for k, v := range config.Terraform.Providers {
-		providers[k] = []map[string]any{}
+		providers[k] = []*ProviderConfigProperties{}
+
 		for _, provider := range v {
-			providers[k] = append(providers[k], provider.AdditionalProperties)
+			var convertedSecrets map[string]*SecretReference
```

Comment:
we are using `providers` above and not `convertedProviders`, can we use same convention here and just name it `secrets`?

Reply:
function is refactored and names updated. pls take another look.